### PR TITLE
Apply PU ID SF only to jets with pT<50

### PR DESCRIPTION
**Pull request by kondratyevd**
As Raffaele mentioned yesterday, the reweighting should be done only for pT<50.
What do you think?

### CODE REVIEW

**Comment by nanlu06** [comment]
Hi Dmitry! 

Thank you very much for making this P.R. I agree. Could you also add the PU ID SF to the dnn vars so that we can look at them when needed?

**Comment by kondratyevd** [comment]
Hi @nanlu06 , thanks for checking! I have added the puid weight to dnn_vars. 
I will also need to change the reference value for tests so the builds don't fail – I will do it after the PR is ready to merge.

**Comment by kondratyevd** [comment]
Actually, I'm still not sure what caused the jets with pT>50 to get affected - the provided efficiency and SF maps only have values for pT bins up to 50 GeV.
@jpata , do you happen to know if that is a normal behavior for `extractor()` , or a bug in Coffea, or something else?

**Comment by jpata** [comment]
Hi, I haven't used that functionality in coffea, it would be wise to check the outputs. What I could suggest to do is set the eff value to 1.0 for jets outside the range using a simple array mask.

**Comment by kondratyevd** [comment]
On a second thought, it makes sense, because the event weights are constructed using the efficiencies of all jets in an event. So even if a particular jet has pT>50GeV, other jets in the event might be softer and thus affect the event weight. So it shouldn't be surprising when jet pT distributions are affected beyond 50 GeV. 
In any case, in this PR I made sure that jets >50GeV are not used when the event weights are calculated.
Sorry for the confusion!

**Comment by kondratyevd** [comment]
I think it's ready to merge, unless there are other comments.

File: tests/hmm/hmumu_utils.py [hunk 1]
```diff
@@ -1542,17 +1544,18 @@ def jet_puid_evaluate(evaluator, era, wp, jet_pt, jet_eta):
...
     for iev in numba.prange(len(offsets)-1):
         p_tot = 1.0
         #loop over jets in event
         for ij in range(offsets[iev], offsets[iev+1]):
-            this_jet_passes = jets_mask_passes_id[ij]
-            if this_jet_passes:
-                p_tot *= jets_eff[ij]
-            else:
-                p_tot *= 1.0 - jets_eff[ij]
+            if jet_pt[ij] < 50:
```
Comment (jpata) [inline tests/hmm/hmumu_utils.py]: Would it be OK to put this cut of 50 GeV somewhere configurable? For example, propagating it from here: https://github.com/jpata/hepaccelerate-cms/blob/master/tests/hmm/pars.py#L499. The issue is that such things hidden deep in the code can easily get forgotton, and it's better to have them in a single place. Thank you!

**Comment by jpata** [comment]
@kondratyevd, we've noticed some issues still with the PUID weight for e.g. 2018. going forward, let's provide a full set of data/mc validation plots for any non-trivial PR, see e.g. here: https://github.com/jpata/hepaccelerate-cms/pull/57#issuecomment-555309851

Can you please do that?

**Comment by kondratyevd** [comment]
@jpata, I can currently quickly produce full plots only for 2016. 
For 2017 and 2018 I can do it only with nanoAODv5 (no fsr) so far, I'm not sure if copying and processing nanoAODv6 will be ready this week. Let me know what you prefer.

**Comment by jpata** [comment]
If you can only validate for 2016, how about circulating (again) the plots for 2016 with PUID off/on and disabling the weight for the other years?

**Comment by kondratyevd** [comment]
@jpata I haven't figured out yet how to set up a website at web.cern.ch, but for now all plots for 2016 are here:
https://www.dropbox.com/sh/gv8deb4gwhfn7i8/AACfD801h7M347JHpApI6Mfka?dl=0

**Comment by jpata** [comment]
Thanks Dmitry! I agree it looks fine and improves the agreement for 2016. As we didn't test for 2017 or 2018, I would propose to disable the puid weight there. Would you add a simple `and dataset_era == "2016"`, with a comment referencing this discussion here? Also, would you mind to fix the conflicts? Thank you!

**Comment by kondratyevd** [comment]
Hi @jpata , I agree, disabling the weights for 2017/18 is a good idea. I'll try to look into 2017/18 performance soon.

I have resolved the conflicts, so the PR can be merged, unless there are other comments

**Comment by jpata** [comment]
Thank you! Let me run a quick validation also on this end, and if all looks good, I will merge it today. Just trying to prevent another data/mc divergence due to many changes that we didn't test end-to-end...

**Comment by jpata** [comment]
For 2017-2018, it crashes with:
```python
Traceback (most recent call last):
  File "tests/hmm/analysis_hmumu.py", line 584, in <module>
    main(args)
  File "tests/hmm/analysis_hmumu.py", line 561, in main
    run_analysis(args, outpath_partial, jobfile_data, analysis_parameters, analysis_corrections)
  File "/storage/user/jpata/hmm/dev2/hepaccelerate-cms/tests/hmm/hmumu_utils.py", line 1058, in run_analysis
    lumimask = analysis_corrections.lumimask)
  File "/storage/user/jpata/hmm/dev2/hepaccelerate-cms/tests/hmm/hmumu_utils.py", line 1160, in event_loop
    **kwargs)
  File "/storage/user/jpata/hmm/dev2/hepaccelerate-cms/hepaccelerate/hepaccelerate/utils.py", line 530, in analyze
    ret = analyze_data(data, **kwargs)
  File "/storage/user/jpata/hmm/dev2/hepaccelerate-cms/tests/hmm/hmumu_utils.py", line 559, in analyze_data
    dnn_vars["puid_weight"] = weights_individual['jet_puid']['nominal'][dnn_presel]
KeyError: 'jet_puid'
```

**Comment by jpata** [comment]
How about setting the puid weight to 1 for 2017 and 2018, making a note that this will be removed once it has been tested for those years?

**Comment by kondratyevd** [comment]
Sure, that would work as well. Could you test on 2017 or 18 now?

**Comment by jpata** [comment]
Thanks, the jobs are running!

**Comment by jpata** [comment]
Validation plots with full systematics: https://jpata.web.cern.ch/jpata/hmm/results/d7b1bf9629f85f246db24009a7e689e505826d24/
- number of jobs: 709
- runtime = 960.94 +- 671.09 s
- maxrss = 2360.45 +- 209.85 MB
- speeds = 1884.41 +- 1616.63 Hz

https://jpata.web.cern.ch/jpata/hmm/results/d7b1bf9629f85f246db24009a7e689e505826d24/plots/2016/pdf/baseline_hist__dimuon__leading_muon_pt_nominal.pdf
https://jpata.web.cern.ch/jpata/hmm/results/d7b1bf9629f85f246db24009a7e689e505826d24/plots/2017/pdf/baseline_hist__dimuon__leading_muon_pt_nominal.pdf
https://jpata.web.cern.ch/jpata/hmm/results/d7b1bf9629f85f246db24009a7e689e505826d24/plots/2018/pdf/baseline_hist__dimuon__leading_muon_pt_nominal.pdf

From my point of view, looks OK to merge. @irenedutta23 @nanlu06 do you want to run a check on the limits, just to make sure there are no surprises there?

**Comment by irenedutta23** [comment]
Do you apply a jet pt > 25 requirement before calculating the event weight? If not, you will take more jets into account than just those that are actually passing our object selection.
https://github.com/jpata/hepaccelerate-cms/pull/66/files#diff-7c681c3161b87061604d79dc8695e4ffR336
I don't see that requirement anywhere in the function

For reference, see this line in the btag weight calculation, where we set jet sf =1. if the jet isn't within the eta boundaries and doesn't pass the pt selection.
https://github.com/jpata/hepaccelerate-cms/pull/67/files#diff-7c681c3161b87061604d79dc8695e4ffR1696-R1713

**Comment by kondratyevd** [comment]
Hi @irenedutta23, no, I was under impression that `selected_jets` come already with all selections applied. I can see now that they come with eta cut but without pT cut.

I will fix it soon, hopefully it will make Data/MC agreement better.

**Comment by jpata** [comment]
Thanks Irene for checking the code! To clarify, jet pt cuts are only applied in the JES/JER systematic loop on https://github.com/jpata/hepaccelerate-cms/blob/master/tests/hmm/hmumu_utils.py#L431. If you depend on cuts on the nominal jet pt for weight calculation, it should be applied separately.

**Comment by kondratyevd** [comment]
Okay, I think the fix is done. 
@jpata, could you reproduce the validation plot for 2016? 

I have the copies of almost all of 2018 v6 datasets, so I'll be able to enable the weights and take a look at Data/MC for 2018 soon.

**Comment by kondratyevd** [comment]
Here are the plots for 2018 nanoAODv6:
w/o puid weights: https://www.dropbox.com/sh/kz7wm73yzubdbwj/AAAMQlzCJiKyx94lS1NqQOhWa?dl=0
with puid weights: https://www.dropbox.com/sh/dh9q21jh2tnofpp/AAAjs2U-1kGD5kzz1mYCLFW8a?dl=0

In these plots `dy_m105_160_mg` and `dy_m105_160_vbf_mg` samples are used instead of corresponding aMC samples, because `dy_m105_160_amc` is not yet available for 2018.

**Comment by nanlu06** [comment]
Hi Dmitry! Thank you very much for the 2018 validation plots. It looks nice to me.

**Comment by jpata** [comment]
Thanks for the effort! Let me rerun on my side all the data/mc plots, I will post them here.

**Comment by jpata** [comment]
https://jpata.web.cern.ch/jpata/hmm/results/64ee178a851d30765db0d62c9cc3eab6cd789066

- number of jobs: 709
- runtime = 1045.58 +- 918.24 s
- maxrss = 2361.66 +- 217.35 MB
- speeds = 1882.95 +- 1759.28 Hz

From my point of view, looks OK. @nanlu06, @irenedutta23 please check as well that you are satisfied with the control plots. Unless you see issues, I will merge this today in the next hours.

**Comment by nanlu06** [comment]
Thank you Joosep! They look fine to me, too.
________________________________

**Pull request merged**